### ISSUE TITLE
[BUG] Delays on distributed calls to free agents in Call-center

### ISSUE THREAD

**Issue by MarLarTD** [OpenSIPS/opensips#2679]
**OpenSIPS version you are running**

```
version: opensips 3.2.3 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on 00:53:09 Oct 21 2021 with gcc 4.8.5

Also tested with version 3.1.1, 3.1.6, same issue. 
```

**Describe the bug**

When all agent is “in call” the “wrap up” status is not ended to “free” state on first agent that is free, but from last one. This cause a huge delay on when call is distributed to the free agent. 

Scenario:
10 calls is made to call center. We have 10 free agents. All agent takes the calls and all 10 agents is now in “in call” state. Another call is made and is put in queue (MOH played all the time). 
Let us name agent for Agent 1, Agent 2, Agent 3 and so on. Agent 1 ends the call first, say at 08:01:00. Almost the same time call on agent 2 (08:01:01), agent 3 (08:01:02), agent 4 (08:01:03) is ended and so on. When running “opensips-cli -x mi cc_list_agents” you will see that Agent 1 is now in state “wrap up”. You can also see that Agent 2, 3, 4, 5 and so on is in state “wrap up”. When it comes to agent 10, the agent that received the call last and ends the call last it is also put in “wrap up” state. Problem here is that agent 1 is still in “wrap up” state (according to cc_list_agents). After 1 second (default wrapup time on all agents) agent 10 is moved to “free” state at say 08:01:09. Now we can see the issue. Agent 9 is then changed from “wrap up” to “free” state, Agent 8 is then changed from “wrap up” to “free” state and so on. Even that agent1-9 has been free for longer time than agent 10 it starts to free up agent 10 “wrap up” time instead of agent 9-1. 
The call that was put in queue, has for obvious and correct reason been told by module to send call to agent 1 as it has ended the call first (at 08:01:00). But as agent 1 is sent to “free” state 10 seconds (at 08:01:19) after agent 10 has ended the call our call in queue is placed in a queue for up to 10 seconds even that you have 10 free agents. So if you have a scenario of 100 agents, your call will be placed in a queue for 100 seconds even if you have 99 free agents. 1000 agents = 1000 seconds. 

**To Reproduce**

Configure 10 agents in cc_agents. Wrap up time set to 1 second (tested both in config file, on cc_flows and on cc_agents same issue)
This issue can be reproduced using sipp. 

`sipp -i <sipp-host> -s <plus international number> -m 10 -r 1 -rp 100 -d 20000 -sn uac <opensips-host>`
This will put 10 agents in "in call" state for 20 seconds. It will send 1 call every 0.1 second. 
Then you call from your mobile/landline to the opensips-host. This call will be put in queue.  
Then you run "opensips-cli -x mi cc_list_agents" to check "State" value. You will see all is "in call" state. When call is ended you will see how agent state is changed to "wrap up". And you will notice that your call in queue will be sent to the agent when all agent is in state "free". Not before (even if you see 9 other agents as free). 

**Expected behavior**

Agent that ends call first must be marked as "free" first. 

**Relevant System Logs**

DBG log has been enabled, but I do not find any clue in it. When agent is moved from "wrap up" to "free" state it does not log anything in log. 
DBG:call_center: shows when call is setup. But then it floods "WARNING:core:log_expiry: threshold exceeded : msg processing took too long - 21150 us.Source" so it does not tell anything more when call is ended. 
If I do a single call I do see full trace, so I tested the limit of dgb log and it could handle 3 calls to agent. You can reproduce this issue also with 3 agents. 

14:26:58 call is received on agent 1. 14:27:18 call is ended by agent 1. But is still marked as "wrap up" until 14:27:23 (3-4 seconds over 1 second wrap up time). But same time when last agent is marked as free 14:27:23 I see TM doing logging about deleted transaction. Could this give us any hint of the issue?

Below log is filtered on call 0x7f601eb2ce00 and all TM logs when all agent is "free" (according to cc_list_agents). 
```
Nov  9 14:26:58 q02 opensips[22689]: DBG:call_center:print_call_list: [] - 0x7f601eb2ce00
Nov  9 14:26:58 q02 opensips[22689]: DBG:call_center:w_handle_call: ETA for new call(0x7f601eb2ce00) is 0
Nov  9 14:26:58 q02 opensips[22689]: DBG:call_center:w_handle_call: new destination for call(0x7f601eb2ce00) is sip:+xxx@xxx:5060 (state=6)
Nov  9 14:26:58 q02 opensips[22689]: DBG:call_center:set_call_leg: call 0x7f601eb2ce00 moving to sip:+xxx@xxx:5060 , state 6
Nov  9 14:26:58 q02 opensips[22713]: DBG:call_center:print_call_list: [949.0] - 0x7f601eb2ce00
Nov  9 14:26:58 q02 opensips[22713]: DBG:call_center:print_call_list: [949.0] - 0x7f601eb2ce00
Nov  9 14:27:18 q02 opensips[22689]: DBG:call_center:b2bl_callback_customer:  call (0x7f601eb2ce00) has event 1,
Nov  9 14:27:18 q02 opensips[22689]: DBG:call_center:terminate_call: terminating call 0x7f601eb2ce00 (stat=0x7fffe5140c20)
Nov  9 14:27:18 q02 opensips[22713]: DBG:call_center:b2bl_callback_customer:  call (0x7f601eb2ce00) has event 4,
Nov  9 14:27:18 q02 opensips[22713]: DBG:call_center:b2bl_callback_customer: A delete in b2blogic, call->state=7, 0x7f601eb2ce00
Nov  9 14:27:18 q02 opensips[22713]: DBG:call_center:free_cc_call: free call 0x7f601eb2ce00, [949.0]

Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:timer_routine: timer routine:2,tl=0x7f601eb87ce8 next=0x7f601eb86280, timeout=3296
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: removing 0x7f601eb87c68 from table
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:delete_cell: delete transaction 0x7f601eb87c68
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: done
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:timer_routine: timer routine:2,tl=0x7f601eb86280 next=0x7f601eb5ae18, timeout=3296
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: removing 0x7f601eb86200 from table
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:delete_cell: delete transaction 0x7f601eb86200
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: done
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:timer_routine: timer routine:2,tl=0x7f601eb5ae18 next=0x7f601eb593b0, timeout=3296
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: removing 0x7f601eb5ad98 from table
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:delete_cell: delete transaction 0x7f601eb5ad98
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: done
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:timer_routine: timer routine:2,tl=0x7f601eb593b0 next=0x7f601eb946b0, timeout=3296
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: removing 0x7f601eb59330 from table
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:delete_cell: delete transaction 0x7f601eb59330
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: done
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:timer_routine: timer routine:2,tl=0x7f601eb946b0 next=0x7f601eb8fa68, timeout=3296
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: removing 0x7f601eb94630 from table
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:delete_cell: delete transaction 0x7f601eb94630
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: done
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:timer_routine: timer routine:2,tl=0x7f601eb8fa68 next=(nil), timeout=3296
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: removing 0x7f601eb8f9e8 from table
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:delete_cell: delete transaction 0x7f601eb8f9e8
Nov  9 14:27:23 q02 opensips[22697]: DBG:tm:wait_handler: done
```

**OS/environment information**
 - Operating System: CentOS Linux release 7.9.2009 (Core)
 - OpenSIPS installation: yum (for 3.1.x) and rpm install (upgrade from 3.1.x to 3.2.3)
 - other relevant information: SELinux enabled, IPtables enabled 

**Additional context**

What I have tested
* Disabled SELinux
* Disabled IPTables
* 3.1.1 version
* 3.1.6 version
* Increased/Decreased S_MEMORY, as moment I use 2048
* Increased/Decreased P_MEMORY, as moment I use 32
* Tested timer_workers 1 and 3, as moment I use 3
* Increased/Decreased udp_worksers, as moment I use 32
* Postgres on standalone host and localhost 
* Disabled all modules not needed by call-center module
* Made a very tiny route that basically only did "if (!cc_handle_call("flow_id"))"
* Tested various wrap up timers (config, db cc_flows, db cc_agents), if i increase wrap_up time also the wrap_up time is extended in scenario, so the wrap_up time period itself seems to work fine. 
* Tested b2b cleanup_period
* Tested some TM timers, but perhaps I should test something more here?
* Tested to write B2B to db and not

I do not see any memory issue or CPU issue on server. 
"opensips-cli -x diagnose memory" and "opensips-cli -x mi get_statistics shmem: core:" seems good, dont see anything wrong with memory. 

Also unix timestamp in db is correct.

**Comment by MarLarTD** [comment]
I found a workaround (not nice one) that you every 5 sec hangup the call on MOH server (I am generating a ringback tone). It will generate a BYE from MOH/media server to opensips. Opensips will generate a new in-dialog invite, If still queue, it will send it back to MOH server (that will send BYE again after 5 sec). But if free agents, it will send call to the free agent directly even if you see this delay of "wrap up" of other agents. So it will be like a check that every 5 sec checks for free agents. As I run only ringback tone our customer will not notice anything. 

However, this will limitation the use of queue statistics as it will generate a new call to it every 5 sec. But for me this will be a workaround I will need to run on production to get rid of this delay until we have found the reason of this delay.
